Foreign fields pt 2 (#777)
o1-labs/docs2: docs/zkapps/o1js/foreign-fields.mdx

y.toBigInt() === 5n; // convert to bigint
```

As usual, you can find more information about each method in the [API reference](../o1js-reference/classes/ForeignField).

> ## Three kinds of foreign field

Reviewer: ```suggestion
## Three kinds of foreign fields
```

If the basic usage examples look straightforward, here is where it gets a bit complicated.

> For each `ForeignField` class created with `createForeignField()`, there are actually three different variants. We call them _unreduced_, _almost reduced_ and _canonical_.

Reviewer: ```suggestion
For each `ForeignField` class created with `createForeignField()`, there are actually three different variants: _unreduced_, _almost reduced_ and _canonical_.
```

You find the variants as static properties on the class; they are themselves classes:

```ts
let x = new Field17.Unreduced(0);
let y = new Field17.AlmostReduced(0);
let z = new Field17.Canonical(0);
```

Unreduced field elements just have the `ForeignField` type. For the other two variants, there are narrower base types that are common to each variant:

```ts
import { AlmostReducedField, CanonicalField } from 'o1js';

y satisfies AlmostReducedField;
z satisfies CanonicalField;
```

> In the following, we explain when to use the different variants, and how to convert between them. You don't need to learn all of it, though: The type system will guide you to use the right variant in each situation.

Reviewer: ```suggestion
In the following section, you learn when to use the different variants and how to convert between them. You don't need to learn all of it, though: The type system guides you to use the right variant in each situation.
```

Author: ```suggestion
In the following section, you learn when to use the different variants, and how to convert between them. You don't need to remember all of it, though: The type system guides you to use the right variant in each situation.
```

### Unreduced fields

Most arithmetic operations return unreduced fields:

```ts
import assert from 'assert';

let z = x.add(x);
assert(z instanceof Field17.Unreduced);
```

In short, **unreduced** means that a value can be larger than the modulus.

For example, if `x` has the value 16, it is valid for `x.add(x)` to contain the value 32. The addition is correct modulo 17, but doesn't guarantee a result smaller than 17.

:::note

> Unreduced doesn't usually mean that the underlying witness is larger than the modulus. It just means that we have not _proved_ it to be smaller.
> A malicious prover _could_ make it larger, by slightly modifying their local version of o1js and creating a proof with that.

Reviewer: ```suggestion
Unreduced doesn't usually mean that the underlying witness is larger than the modulus. It just means that it is not _proved_ to be smaller.
A malicious prover _could_ make it larger by slightly modifying their local version of o1js and creating a proof with that version.
```

:::

Unreduced fields can be added and subtracted, but not multiplied or divided:

```ts
z.add(1).sub(x); // works

assert((z as any).mul === undefined); // z.mul() is not defined
assert((z as any).inv === undefined);
assert((z as any).div === undefined);
```

### Almost reduced fields

To do multiplication, you need almost reduced fields. You can convert to them by using `.assertAlmostReduced()`:

```ts
let zAlmost = z.assertAlmostReduced();
assert(zAlmost instanceof SmallField.AlmostReduced);
```

Now you can do multiplication and division:

```ts
let zz = zAlmost.mul(zAlmost); // zAlmost.mul() is defined

// but .mul() returns an unreduced field again:
assert(zz instanceof SmallField.Unreduced);

// zAlmost.inv() is defined, and returns an almost reduced field:
assert(zAlmost.inv() instanceof SmallField.AlmostReduced);
```

> It can be convenient to require almost reduced fields as inputs to your smart contract. To do so, create a class which can also serve as a type, and use its `.provable` property when passing to the state decorator:

Reviewer: ```suggestion
It can be convenient to require almost reduced fields as inputs to your smart contract. To do that, create a class that can also serve as a type and use its `.provable` property when passing to the state decorator:
```

```ts
class AlmostField17 extends Field17.AlmostReduced {}

class MyContract extends SmartContract {
  @state(AlmostField17.provable) x = State<AlmostField17>();

  @method myMethod(y: AlmostField17) {
    let x = y.mul(2);
    this.x.set(x.assertAlmostReduced());
  }
}
```

#### What does almost reduced mean?

> The definition of almost reduced is somewhat technical. The main motivation is to guarantee that the way we prove modular multiplication is sound. That is definitely true for field elements `< 2^259`. (Recall that we require the modulus to be `< 2^259`.)
> 
> However, we actually prove a stronger condition, which lets us save a few constraints in some places:

Reviewer: ```suggestion
The definition of almost reduced is somewhat technical. The main motivation is to guarantee that the way you prove modular multiplication is sound. That is definitely true for field elements `< 2^259`. (Recall that the modulus is required to be `< 2^259`.)

However, we actually prove a stronger condition, which lets us save a few constraints in some places:
```

`z` is **almost reduced** modulo `f`, if `z >> 176` is smaller or equal than `f >> 176`. (`>>` means a [right shift](https://en.wikipedia.org/wiki/Arithmetic_shift).)

:::note

Example: Assume `x` is a `UInt256` holding the value `2^130`. After computing `z = x.mul(x)`, it is valid for `z` to be `2^260`.

> However, by calling `z.assertAlmostReduced()`, we prove that `z` is smaller than `2^259` and safe to use in another multiplication. According to our stronger definition, we even have `z < 2^256`.
> 
> :::
> 
> Why are we exposing `AlmostReducedField` as a separate type, and don't _always_ prove conditions necessary for multiplication? Because that would take up additional constraints!

Reviewer: ```suggestion
However, by calling `z.assertAlmostReduced()`, you prove that `z` is smaller than `2^259` and is safe to use in another multiplication. According to this stronger definition, you even have `z < 2^256`.

:::

Why is `AlmostReducedField` exposed as a separate type, and why is _always_ prove conditions required for multiplication? Because that would take up additional constraints!
```

Author: This is a tough one for me, because "we" is so standard in the mathematical arguments I'm used to writing

```suggestion
However, by calling `z.assertAlmostReduced()`, you prove that `z` is smaller than `2^259` and safe to use in another multiplication. According to the stronger definition, you even have `z < 2^256`.

:::

Why is `AlmostReducedField` exposed as a separate type, instead of _always_ proving conditions necessary for multiplication? Because that would take up additional constraints!
```

Author: "necessary condition" is standard math language so I'm going to leave that

> `ForeignField` is built to allow you to use the minimum amount of constraints, in a way that is safely guided by the type system. See the section below on [minimizing constraints](#minimizing-constraints) for more details.

Reviewer: ```suggestion
`ForeignField` is built to allow you to use the minimum amount of constraints in a way that is safely guided by the type system. See [minimizing constraints](#minimizing-constraints) for more details.
```

When you create fields from constants, they always get fully reduced. The type signature of `ForeignField.from()` reflects this and returns a canonical field:

```ts
let constant = Field17.from(16);
assert(constant instanceof Field17.Canonical);

// these also work, because `from()` takes the input mod 17:
Field17.from(100000000n) satisfies CanonicalForeignField;
Field17.from(-1) satisfies CanonicalForeignField;
```

You can convert any field to canonical by calling `.assertCanonical()`:

```ts
let zCanonical = z.assertCanonical();
assert(zCanonical instanceof Field17.Canonical);
```

Canonical fields are a special case of almost reduced fields at the type level:

```ts
constant satisfies AlmostForeignField;
constant.mul(constant); // works
```

The cheapest way to prove that an existing field element is canonical is to show that it is equal to a constant:

```ts
let zCanonical = z.assertEquals(3);
assert(zCanonical instanceof Field17.Canonical);
```

An operation that is only possible on canonical fields is the boolean equality check:

```ts
let xCanonical = x.assertCanonical();
let yCanonical = y.assertCanonical();
let isEqual = xCanonical.equals(yCanonical);
```

> We require inputs to be canonical for this because we check for strict equality, not equality modulo the field size. Note that being strictly unequal does not imply being unequal as field elements, so `equals()` on non-canonical fields would be error-prone.

Reviewer: ```suggestion
Inputs must be canonical for boolean equality check operations because the operation checks for strict equality, not equality modulo the field size. Note that being strictly unequal does not imply being unequal as field elements, so `equals()` on non-canonical fields would be error-prone.
```

Author: ```suggestion
Inputs must be canonical for  `equals()` because the operation checks for strict equality, not equality modulo the field size. Note that being strictly unequal does not imply being unequal as field elements, so `equals()` on non-canonical fields would be error-prone.
```
Trying to be a bit less verbose

## Minimizing constraints

#### `assertAlmostReduced()`

Here is a trick to save constraints when you need to "almost reduce" many field elements: Always reduce them in _batches of 3_.

For example, do this when doing many multiplications in a row:

```ts
let z1 = x.mul(7);
let z2 = x.add(11);
let z3 = x.sub(13);

let [z1r, z2r, z3r] = Field17.assertAlmostReduced(z1, z2, z3);

z1r.mul(z2r);
z2r.div(z3r);
```

`assertAlmostReduced()` takes any number of inputs, but is by far the most efficient with multiples of 3. For example:

- 1 input takes 4.5 constraints
- 2 inputs take 5 constraints
- 3 inputs take 5.5 constraints

#### `sum()`

> Another opportunity to save constraints is when doing many additions or subtractions in a row. Instead of doing something like `x.add(y).sub(z)`, use `ForeignField.sum()`:

Reviewer: ```suggestion
Another opportunity to save constraints is when many additions or subtractions are performed in a row. Instead of doing something like `x.add(y).sub(z)`, use `ForeignField.sum()`:
```
